xla/service/gpu/gpu_all_gather_optimizer.cc

```diff
@@ -0,0 +1,194 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/gpu/gpu_all_gather_optimizer.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_casting_utils.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instruction.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instructions.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_module.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_opcode.h"
+#include "tensorflow/compiler/xla/service/collective_ops_utils.h"
+#include "tensorflow/compiler/xla/service/hlo_query.h"
+#include "tensorflow/compiler/xla/service/reduce_scatter_utils.h"
+
+namespace xla {
+namespace gpu {
+
+// Structure to keep track of the subgraph to optimize
+// "Left" and "Right" are conceptual. The initial
+// reduce scatter is always considered "left" and "right"
+// points to the other side of subgraph ending in a binary op
+
+struct AllGatherReduceScatterSpec {
+  HloInstruction* all_gather_left;
+  HloInstruction* all_gather_right;
+  HloInstruction* right_source;
+  HloInstruction* binary_op;
+};
+
+std::optional<AllGatherReduceScatterSpec> MatchReduceScatter(
+    const HloReduceScatterInstruction* rs, int64_t min_rank = 1) {
+  if (!rs->shape().IsArray() || rs->constrain_layout() ||
```

Comment:
The checks here from line 43 to line 67 look like they can be moved to the (to be implemented) InstructionMatchesPattern function, as further code in MatchReduceScatter does not need it.

Reply:
please see the above comment.